File: backTracking/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1:
            return [nums]
        
        resultFinal = []
        for i in range(len(nums)):
            for res in self.permute(nums[:i] + nums[i+1:]):
                resultFinal.append([nums[i]] + res)
        return resultFinal

File: backTracking/permutations.cpp
class Solution {
public:
    vector<vector<int>> res;
    
    void bt(vector<int>& nums, vector<int>& cur_perm, vector<bool>& visited){
        //base cases
        if(cur_perm.size() == nums.size()){
            res.push_back(cur_perm);
            return;
        }

        //recursive cases
        for(int i = 0; i < nums.size(); i++){
            if(!visited[i]){
                cur_perm.push_back(nums[i]);
                visited[i] = true;
                bt(nums, cur_perm, visited);
                cur_perm.pop_back();
                visited[i] = false;
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> cur_perm = {};
        vector<bool> visited(nums.size(), false);
        bt(nums, cur_perm, visited);
        return res;
    }
};
